audit/audit.go

```diff
@@ -0,0 +1,179 @@
+}
+
+type avroMarshaller func(s interface{}) ([]byte, error)
+
+// OutboundProducer defines a producer for sending outbound audit events to a kafka topic
+type OutboundProducer interface {
+	Output() chan []byte
+}
+
+//AuditorService defines the behaviour of an auditor
+type AuditorService interface {
+	Record(ctx context.Context, action string, result string, params common.Params) error
+}
+
+//Auditor provides functions for interception HTTP requests and populating the context with a base audit event and
+// recording audit events
+type Auditor struct {
+	service       string
+	marshalToAvro avroMarshaller
+	producer      OutboundProducer
+}
+
+// NopAuditor is an no op implementation of the AuditorService.
+type NopAuditor struct{}
+
+//Record is a no op implementation of Auditor.Record
+func (a *NopAuditor) Record(ctx context.Context, attemptedAction string, actionResult string, params common.Params) error {
+	return nil
+}
+
+//New creates a new Auditor with the namespace, producer and token provided.
+func New(producer OutboundProducer, namespace string) *Auditor {
+	log.Debug("auditing enabled for service", log.Data{"service": namespace})
+	return &Auditor{
+		producer:      producer,
+		service:       namespace,
+		marshalToAvro: EventSchema.Marshal,
+	}
+}
+
+//Record captures the provided action, result and parameters and an audit event. Common fields - time, user, service
+// are added automatically. An error is returned if there is a problem recording the event it is up to the caller to
+// decide what do with the error in these cases.
+func (a *Auditor) Record(ctx context.Context, attemptedAction string, actionResult string, params common.Params) error {
+	//NOTE: for now we are only auditing user actions - this may be subject to change
+	user := identity.User(ctx)
+	if user == "" {
+		log.Debug("not user attempted action: skipping audit event", nil)
+		return nil
+	}
+
+	if attemptedAction == "" {
+		return NewAuditError("attemptedAction required but was empty", "", actionResult, params)
+	}
+	if actionResult == "" {
+		return NewAuditError("actionResult required but was empty", attemptedAction, "", params)
+	}
+
+	e := Event{
+		Service:         a.service,
+		User:            user,
+		AttemptedAction: attemptedAction,
+		ActionResult:    actionResult,
+		Created:         time.Now().String(),
+		Params:          params,
+	}
+
+	e.RequestID = requestID.Get(ctx)
+
+	avroBytes, err := a.marshalToAvro(e)
+	if err != nil {
+		log.Error(err, nil)
+		return NewAuditError("error marshalling event to arvo", attemptedAction, actionResult, params)
+	}
+
+	log.Info("logging audit message", log.Data{"auditEvent": e})
+	a.producer.Output() <- avroBytes
+	return nil
+}
+
+//NewAuditError creates new audit.Error with default field values where necessary and orders the params alphabetically.
+func NewAuditError(cause string, attemptedAction string, actionResult string, params common.Params) Error {
+	if cause == "" {
+		cause = nilStr
```

Comment:
Why do the default values need to be declared as "nil" rather than just left blank? What would the harm of just calling `Error{"this is a cause", "actionname", "", params}` rather than using the function to create the error be? Should this be documented for the function?

Reply:
Fair point I will update to ""